crates/core/src/metadata_store/providers/objstore/optimistic_store.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use bytestring::ByteString;
+
+use crate::metadata_store::providers::objstore::version_repository::VersionRepositoryError::PreconditionFailed;
+use crate::metadata_store::providers::objstore::version_repository::{
+    TaggedValue, VersionRepository, VersionRepositoryError,
+};
+use crate::metadata_store::{Precondition, ReadError, VersionedValue, WriteError};
+use restate_types::config::MetadataStoreClient;
+use restate_types::Version;
+
+pub(crate) struct OptimisticLockingMetadataStoreBuilder {
+    pub(crate) version_repository: Arc<dyn VersionRepository>,
+    pub(crate) configuration: MetadataStoreClient,
+}
+
+impl OptimisticLockingMetadataStoreBuilder {
+    pub(crate) async fn build(&self) -> anyhow::Result<OptimisticLockingMetadataStore> {
+        let MetadataStoreClient::ObjectStore { .. } = self.configuration else {
+            anyhow::bail!("unexpected configuration value");
+        };
+        Ok(OptimisticLockingMetadataStore::new(
+            self.version_repository.clone(),
+        ))
+    }
+}
+
+pub struct OptimisticLockingMetadataStore {
+    version_repository: Arc<dyn VersionRepository>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
+#[serde(tag = "version", content = "value")]
+enum OnDiskValue<'a> {
+    V1(Cow<'a, VersionedValue>),
+}
+
+fn versioned_value_to_bytes(v: &VersionedValue) -> anyhow::Result<Bytes> {
+    let on_disk = OnDiskValue::V1(Cow::Borrowed(v));
+
+    let mut buf = vec![];
```

Comment:
If we pass in this buffer (using a `BytesMut`), then it could live in the `OptimisticLockingMetadataStore` and be reused across multiple serializations.